### ISSUE TITLE
Memory allocation issue

### ISSUE THREAD

**Issue by oliviertassinari** [mui/mui-toolpad#2340]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

Steps:

1. Deploy a basic toolpad app in an environment with 512 MB of memory
2. Limit the max old space of Node.js to 396 MB `NODE_OPTIONS='--max-old-space-size=396'`

### Current behavior 😯

Crash memory allocation issue

### Expected behavior 🤔

Works

### Context 🔦

Trying to deploy the example of Toolpad in render.com

**Comment by apedroferreira** [comment]
So I'm guessing something like `--optimize_for_size --max_old_space_size=460 --gc_interval=100` won't help?

**Comment by Janpot** [comment]
Running some tests in docker

1. `docker run --rm -it -m 512m -w /app -v $PWD/examples/qr-generator:/app node sh`
2. `yarn`
3. `yarn build`
4. this fails due to memory issues
5. `NODE_OPTIONS='--max-old-space-size=396' yarn build`
6. fails due to memory issues as well
7. `NODE_OPTIONS='--max-old-space-size=450' yarn build`
8. this passes and builds the artifact

Let's try again without limits:
1. `docker run --rm -it -w /app -v $PWD/examples/qr-generator:/app node sh`
2. `yarn build`
3. this passes and builds the artifact

Let's try to run that build
1. `docker run --rm -it -m 512m -w /app -v $PWD/examples/qr-generator:/app node sh`
2. `yarn start`
3. this runs comfortably at around 311M memory usage

Memory issues seem to only happen at build, which is sort of expected to take quite some memory. It seems possible to build it in a container of 512m, if you tune it well. 512m is really not enough though, would love to see anyone attempt at building the docs website under those limitations.

**Comment by oliviertassinari** [comment]
In render.com, it fails at the `yarn start` step. I haven't seen it fail during `yarn build`. Could it be https://github.com/prisma/prisma/issues/4972?

**Comment by Janpot** [comment]
> Could it be https://github.com/prisma/prisma/issues/4972?

Perhaps, if the end-user application is using prisma. Toolpad doesn't use prisma. 

> In render.com, I believe fails at the `yarn start` step. 

I know some things that can be optimized in the function runtime. Will take another look at it.

**Comment by oliviertassinari** [comment]
I have narrowed down a bit more the problem:

- The PR preview on https://github.com/oliviertassinari/mui-olivier-tools/pull/1 breaks. I'm using a free plan
- I created a reproduction a of the issue: https://dashboard.render.com/web/srv-ciu77mdiuiedpv7ao8j0.

**Comment by Janpot** [comment]
In https://github.com/oliviertassinari/mui-olivier-tools/pull/1/files, did this start happening between 0.1.19 -> 0.1.20 or between 0.1.13 -> 0.1.20?

**Comment by oliviertassinari** [comment]
The issue seems to have been introduced in v0.18.0, per https://github.com/oliviertassinari/mui-olivier-tools/pull/1/commits/a69b6195ab8df59e02ad2f96e2abfe330f04c9d8.

**Comment by oliviertassinari** [comment]
Great to see improvements in memory consumption. I had a look at the min memory requirement for more complex apps as I was curious:

- Strapi https://docs.strapi.io/dev-docs/deployment#hardware-and-software-requirements 2GB
- Metabase https://www.metabase.com/docs/latest/installation-and-operation/running-metabase-on-elastic-beanstalk 2GB

But we can't compare Toolpad to these products so much, a bit of a different category.

I tried to upgrade in https://github.com/oliviertassinari/mui-olivier-tools/pull/3 to v0.1.22. It failed. In the last commit, I created a dummy express server to better isolate the problem.

It might be an issue with the initial memory allocated during the start of `toolpad start`?

I think https://mui.com/toolpad/tutorials/render-deploy/#creating-a-new-app-on-render could be set more clear, we don't precise the minimum hardware requirements.

**Comment by oliviertassinari** [comment]
> Could it be toolpad build that's failing? One area where memory consumption could have increased is during toolpad build.

I don't think so, I did https://github.com/oliviertassinari/mui-olivier-tools/pull/3/commits/c380e25a5e0601aef87797dbf088c5213945008c precisely to rule out this possibility.

**Comment by oliviertassinari** [comment]
I'm reopening, as in the worse case, I think we should document the minimum runtime requirement like https://docs.strapi.io/dev-docs/deployment#hardware-and-software-requirements, this also connects to https://github.com/mui/mui-toolpad/discussions/2432.

**Comment by oliviertassinari** [comment]
#2689 solved a significant chunk of the issue. Memory has been down significantly:

<img width="568" alt="Screenshot 2023-09-26 at 15 08 09" src="https://github.com/mui/mui-toolpad/assets/3165635/b44ac766-09c0-41ca-bc9c-39bf733409e2">

I could upgrade https://github.com/oliviertassinari/mui-olivier-tools/pull/5 🎉

---

To close this issue, I think it's mostly a matter of:

- Docs: https://github.com/mui/mui-toolpad/issues/2340#issuecomment-1666552254
- Maybe, having a test in the CI to have observability on this
